### DISCUSSION TITLE
help！！！fail to read file

### DISCUSSION THREAD

**Discussion by hzm199822** [libvips/libvips#2067] [Q&A]
filename=/jpg/bg_2.jpg

I use
local image = vips.Image.new_ from_ file(filename)
however
vips/Image_ methods.lua:137 : VipsForeignLoad: "/jpg/bg_ 2.jpg" is not a known file format

![12](https://user-images.githubusercontent.com/68108652/98906828-c2958400-24f8-11eb-89ee-28126ef3a61b.png)
![21](https://user-images.githubusercontent.com/68108652/98906843-c923fb80-24f8-11eb-86cf-643d9ce9e04b.png)

the type(name)is cdata
Explain that the problem is
vips_ foreign_ find_ load

![22](https://user-images.githubusercontent.com/68108652/98906903-e0fb7f80-24f8-11eb-8029-b56b2acd3cb1.png)

So what's the problem?

**Comment by hzm199822** [comment]
![image](https://user-images.githubusercontent.com/68108652/98918350-794d3080-2508-11eb-8d2e-abb57150a934.png)
However, when I execute in a command-line script, there is no error

**Comment by jcupitt** [comment]
It should work. I see:

```lua
#!/usr/bin/luajit

local vips = require "vips"
local image = vips.Image.new_from_file(arg[1])
print("image width = " .. image:width())
```

Then:

```
$ ./try323.lua ~/pics/k2.jpg
image width = 1450
```

**Comment by hzm199822** [comment]
I use Lua driven by nginx instead of using Lua script directly. not #!/usr/bin/luajit

**Comment by jcupitt** [comment]
Sure, but for debugging you can try command-line tests. If they work, then your problem is probably in nginx.

**Comment by hzm199822** [comment]
To tell you the truth, I'm sure that there is no problem with the path. I can read the file normally in the command line, but in the nginx driver, use this function vips.Image.new_ from_ File will report an error. Finally, the problem is located in VIPs_ lib.vips_ foreign_ find_ Load debugging shows that the type of name is CDATA and the value is nil, that is, in local name = VIPs_ lib.vips_ foreign_ find_ The return value of the function in load (file name) is nil, but why is it empty? Before this function, I debugged that the file name is the correct path

**Comment by hzm199822** [comment]
vips.vips_ foreign_ find_ load  

The feeling is here

**Comment by hzm199822** [comment]
![image](https://user-images.githubusercontent.com/68108652/98936894-97bf2600-2520-11eb-8d8d-4aa9785909c2.png)

**Comment by hzm199822** [comment]
function Image.new_from_file(vips_filename, ...)
    local filename = to_string_copy(vips_lib.vips_filename_get_filename(vips_filename))
    local options = to_string_copy(vips_lib.vips_filename_get_options(vips_filename))

    ngx.log(ngx.ERR,filename)
    ngx.log(ngx.ERR,type(filename))

    local name = Image.find_load(filename)
    if name == nil then
        error(verror.get())
    end

    return voperation.call(name, options, filename, unpack { ... })
end

**Comment by jcupitt** [comment]
I guess you need to find at what point the filename vanishes. Sorry, I don't think I can help.

**Comment by hzm199822** [comment]
OK, thank you. I'll think about it again. Can I consider other reading methods. Or does it not support file reading in this way

**Comment by jcupitt** [comment]
It should work fine, I think you just have a small problem with nginx and strings.

Add some more logging and try to discover exactly where your filename is disappearing.

**Comment by hzm199822** [comment]
I found out that VIPs is after this function_ foreign_ find_ Load, but I can't enter this function to debug

**Comment by hzm199822** [comment]
Only according to the source code you provide, you can only find this layer

**Comment by hzm199822** [comment]
![image](https://user-images.githubusercontent.com/68108652/98938381-bfaf8900-2522-11eb-9a23-33d7bc71750f.png)

**Comment by jcupitt** [comment]
You can debug that using gdb -- just put a breakpoint on `vips_foreign_find_load`. You'll need to build libvips with debugging enabled.
